locale/zh-cn/docs/guides/anatomy-of-an-http-transaction.md

```diff
@@ -0,0 +1,353 @@
+
+```javascript
+const server = http.createServer();
+server.on('request', (request, response) => {
+  // the same kind of magic happens here!
+});
+```
+
+当一个 HTTP 到达服务端，node 调用 request 处理程序，并产生一些唾手可得的对象用以处理传输，这些对象就是  `request` 和 `response`。我们马上会讲到。
+
+实际上，为了处理请求，[`listen`][] 方法需要在 `server` 对象上被显式调用。在大多数情况下，你只要把端口号作为参数传入 `listen` 方法中，作为监听端口即可。当然也有一些其它选项，具体可以参考 [API 参考文档][]。
+
+## 方法, 访问地址以及请求头
+
+当处理一个请求时，第一件事你需要做的是看一下这个方法和其访问地址，以此决定你到底采取何种合理的行为。Node 通过把这些行为属性附加到 `request` 对象上，使得我们处理起来相对而言可以轻松一些。
+
+```javascript
+const { method, url } = request;
+```
+> **注意：** `request` 对象是 [`IncomingMessage`][] 的一个实例。
+
+这里的 `method` 总是一个普通的 HTTP 方法动作行为 (verb)，`url` 是指没有服务器协议和
+端口号的完整访问地址。一个典型的访问地址通常意味着包括第三个斜杠以及后面的所有内容。
+
+请求头也不是很难得到，它们也在 `request` 对象里，称为 `headers`。
+
+```javascript
+const { headers } = request;
+const userAgent = headers['user-agent'];
+```
+
+非常重要的一点是：所有的请求头全是小写字母，而不管实际上它们是怎么进行传输的。所以在无论任何
+情况下，解析请求头就得到了简化。
+
+如果一些请求头出现重复，它们的值不是被覆盖，就是通过英文分号进行分割。究竟哪种方式取决于
+具体的信息头。在某些情况下这可能出现问题，所以我们还可以直接使用 [`rawHeaders`][]。
+
+## 请求体
+
+当接受到了一个 `POST` 或者 `PUT` 请求时，请求体对于你的应用程序非常重要。相对于访问请求
+头而言，获取请求体有些麻烦。传入请求对象的 `request` 其实实现了 [`ReadableStream`][] 接口，
+这个信息流可以被监听，或者与其它流进行对接。我们可以通过监听 `'data'` 和 `'end'` 事件从而把
+数据给取出来。
+
+每次在 `'data'` 事件中触发抓获的数据块是一个 [`Buffer`][]。如果你已知是一个字符串对象，那么
+最好的方案就是把这些数据收集到一个数组中，然后在 `'end'` 事件中拼接并且把它转化为字符串。
+
+```javascript
+let body = [];
+request.on('data', (chunk) => {
+  body.push(chunk);
+}).on('end', () => {
+  body = Buffer.concat(body).toString();
+  // at this point, `body` has the entire request body stored in it as a string
+});
+```
+
+> **注意：** 或许你感到这样做有些枯燥乏味，不过大多数情况下你算是幸运的。
+因为 [`npm`][] 上实在有太多的诸如 [`concat-stream`][] 和 [`body`][] 一类类库屏蔽了部分
+细节逻辑而替你做了这些事情。当然，对于你而言在使用这些类库前知道它们到底干了什么非常重要，这
+就是你为什么需要读这篇文章！
+
+## 一笔带过关于错误的一些信息
+
+因为 `request` 是一个 [`ReadableStream`][] 对象，它同样也是 [`EventEmitter`][] 对象。所以当有错误发生时，表现的行为是很相像的。当有错误在 `request` 流上发生时，它会自动激发自身的 `'error'` 事件。**如果你不去处理监听这个事件，此错误将被*抛出*，这导致你的程序崩溃。** 你应该无论如何都要添加  `'error'` 事件去监听你的请求对象，哪怕你只是做一个日志或者用你自己的独有方式去处理（当然，最佳的处理方式是返回一些出错的信息，这已是后话了）。
+
+```javascript
+request.on('error', (err) => {
+  // This prints the error message and stack trace to `stderr`.
+  console.error(err.stack);
+});
+```
+
+当然还有一些其它的方法来 [处理错误][]，诸如其它的抽象化概念和工具等。但是你总是要意识到错误的确会发生，所以你应当处理它们。
+
+## 我们已经聊得那么多了
+
+直到现在，我们已经谈到了如何创建一个对象，如果从请求中获取方法，请求地址，请求头和请求体。当我们把它们组合到一起，它就看上去是这个样子：
+
+```javascript
+const http = require('http');
+
+http.createServer((request, response) => {
+  const { headers, method, url } = request;
+  let body = [];
+  request.on('error', (err) => {
+    console.error(err);
+  }).on('data', (chunk) => {
+    body.push(chunk);
+  }).on('end', () => {
+    body = Buffer.concat(body).toString();
+    // At this point, we have the headers, method, url and body, and can now
+    // do whatever we need to in order to respond to this request.
+  });
+}).listen(8080); // Activates this server, listening on port 8080.
+```
+
+如果我们运行这个示例代码，我们只能 *接收* 到请求但得不到 *回应*。实际上，如果你在浏览器内运行这个示例，你的请求只会超时，因为服务器那边根本没有返回给客户端任何东西。
```

Comment:
No. Sorry but I have to say that I disagree with you on the point.
Notice that this just emphises what we do in a verb——Here `request` and `response` is the verb of the whole sentence instead of a snippet of codes.